lib/environ/src/environ.rs

```diff
@@ -341,17 +359,31 @@ impl<'module_environment> cranelift_wasm::FuncEnvironment for FuncEnvironment<'m
         &mut self,
         mut pos: FuncCursor,
         table_index: TableIndex,
-        _table: ir::Table,
+        table: ir::Table,
         _sig_index: SignatureIndex,
         sig_ref: ir::SigRef,
         callee: ir::Value,
         call_args: &[ir::Value],
     ) -> WasmResult<ir::Inst> {
-        // TODO: Cranelift's call_indirect doesn't implement bounds checking
-        // or signature checking, so we need to implement it ourselves.
         debug_assert_eq!(table_index, 0, "non-default tables not supported yet");
+
+        let callee_ty = pos.func.dfg.value_type(callee);
+        debug_assert_eq!(callee_ty, I32);
+
+        // For some reason table_addr only works with I64.
+        let callee_i64 = pos.ins().uextend(I64, callee);
+        let table_entry_addr = pos.ins().table_addr(I64, table, callee_i64, 0);
```

Comment:
The current limitation is that it be the same type as a pointer, fwiw.

Reply:
Oh, didn't know about that.

Then I would probably go with `pointer_type()` and extend only if needed.